### PR TITLE
Add names to clients

### LINKED ISSUE #93: Each client should set its own name
Right now we're autodiscovering it, and this is fragile. Also unit test each name, since we might use the names as keys in https://github.com/AtomicMaps/secure-assets.

### LINKED ISSUE #234: cli needs a method to specify a client
Once #93 is addressed, it would be good to propagate those names out to the CLI as options, i.e.

```shell
curl -s https://cmr.earthdata.nasa.gov/stac/LPCLOUD/collections/MOD21A1D_061/items/MOD21A1D.A2020253.h08v05.061.2020346022258 \
    | stac-asset download  --client earthdata -f downloaded.json
```

### PR DESCRIPTION
**Pull request by gadomski**
## Related issues and pull requests

- Closes #93 
- Closes #234 
- Follow on issue: https://github.com/stac-utils/stac-asset/issues/240

## Description

Bit a YOLO implementation, but I tested locally by forcing s3:

```shell
$ stac-asset download --client s3 tests/data/item.json 2>&1 | grep s3_client -A 1
  File "/Users/gadomski/Code/stac-utils/stac-asset/src/stac_asset/s3_client.py", line 113, in open_url
    response = await client.get_object(**self._params(url))
```

## Checklist

- [ ] Add tests
- [x] Add docs
- [x] Update CHANGELOG

### CODE REVIEW

File: src/stac_asset/client.py [hunk 1]
```diff
@@ -226,15 +235,35 @@ async def get_client(self, href: str) -> Client:
...
+        EarthdataClient,
+        FilesystemClient,
+        HttpClient,
+        PlanetaryComputerClient,
+        S3Client,
+    ]
+    for client_class in client_classes:
+        if client_class.name == name:
+            return client_class
+    raise ValueError(f"no client with name: {name}")
```
Comment (ircwaves) [inline src/stac_asset/client.py:269]: Needs more oregano, or similar spice to this:

```suggestion
    client_classes: dict[str, type[Client]] = {
        c.name: c for c in Client.__subclasses__()
    }

    try:
        return client_class[name]
    except KeyError:
        raise ValueError(f"no client with name: {name}")
```

File: src/stac_asset/_cli.py [hunk 2]
```diff
@@ -61,6 +61,12 @@ def cli() -> None:
 @cli.command()
 @click.argument("href", required=False)
 @click.argument("directory", required=False)
+@click.option(
+    "-c",
+    "--client",
+    help="Set the client to use for all downloads. If not "
+    "provided, the client will be guessed from the asset href.",
+)
```
Comment (ircwaves) [inline src/stac_asset/_cli.py:69]:
```suggestion
@click.option(
    "-c",
    "--client",
    type=click.Choice([c.name for c in Client.__subclasses__()]),
    help="Set the client to use for all downloads. If not "
    "provided, the client will be guessed from the asset href.",
)
```

**Review by ircwaves — APPROVED** [review]
The two suggestions bring the dynamism I like, that is maybe unnecessary.

**Comment by gadomski** [comment]
Brought in a bit of a hotter spice 🌶️ thanks for the review!

**Pull request merged**